SchemaMigration/COSchemaMigration.h

```diff
@@ -0,0 +1,417 @@
+ * existing libraries depending on it. If a library introduces a breaking change, 
+ * it must be bump its major version to prevent the loading of libraries that
+ * are not compatible with this new schema without an update.
+ *
+ * @section Schema Breaking Changes
+ *
+ * To explain how CoreObject can support breaking changes cleanly, we will 
+ * consider two packages A and B and bundled in two respective libraries AL and 
+ * BL.
+ *
+ * B depends on A and AL depends on BL.
+ *
+ * For a breaking change in A, we must:
+ *
+ * <list>
+ * <item>increment A version and increment AL major version</item>
+ * <item>increment B version and increment BL minor version (or major version if 
+ * we introduce a breaking change to cope with the breaking change in A)</item>
+ * <item>bundle the breaking-change migration targeting A in AL</item>
+ * <item>bundle a migration targeting B in BL to cope with the breaking change 
+ * in A</item>
+ * <list>
+ *
+ * For properties related operations in A, although they are potentially 
+ * breaking changes, in most cases they won't interfer with B, so incrementing B 
+ * package version, and bundling a migration that simply updates the B version 
+ * stored in the items is all what is needed.
+ *
+ * For a breaking change, incrementing the library major version is important to
+ * prevent loading incompatible package versions present in other libraries. 
+ * See Libary Versioning and Metamodel Compatibility section on this topic.
+ *
+ * To understand precisely what is a breaking change, see Supported Migration
+ * Operations section.
+ *
+ * @section Forward Linear Updates
+ *
+ * CoreObject migrations are always applied in a linear fashion to ensure
+ * there is only a single migration code path between two versions (even when 
+ * they are separated by ten intermediate versions). This prevents multiplying
+ * the potential migration paths every time the version is incremented.
+ *
+ * In other words, to migrate from version 2 to 5, the following migrations
+ * occur sequentially:
+ *
+ * <list>
+ * <item>2 to 3</item>
+ * <item>3 to 4</item>
+ * <item>4 to 5</item>
+ * </list>
+ *
+ * Backward migration (e.g. 3 to 2) is unsupported.
+ *
+ * @section Creation and Registration
+ *
+ * The main migration method is -migrateItems:.
+ *
+ * You create a migration by overriding -migrateItems: in a subclass, or setting 
+ * a migration block, finally you must register it at launch time.
+ *
+ * <example>
+ * COSchemaMigration *migration = [COSchemaMigration new];
+ *
+ * migration.packageName = @"com.company.PackageName"
+ * migration.destinationVersion = 3
+ * migration.migrationBlock = ^(COSchemaMigration *migration, NSArray *storeItems)
+ * {
+ *     NSMutableArray *newItems = [NSMutableArray new];
+ *
+ *     for (COItem *item in storeItems)
+ *     {
+ *         newItem = [item mutableCopy];
+ *         [newItem removeValueForAttribute: @"whatever"];
+ *         [newItem setVersion: migration.destinationVersion
```

Comment:
Maybe we should leave it to the migration block for now, otherwise we'd have to make another copy of each item to set the version.